File: python-sdk/observability/tracing/tracer.py
"""Tracer factory for Python services."""

import os
import sys
from typing import Optional

from opentelemetry import trace
from opentelemetry.sdk.resources import Resource, SERVICE_NAME, SERVICE_VERSION
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.sdk.trace.sampling import ParentBasedTraceIdRatio
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter


_tracer_provider: Optional[TracerProvider] = None


def new_tracer(service_name: str) -> trace.Tracer:
    """Create a new OpenTelemetry tracer with OTLP exporter.
    
    Args:
        service_name: The name of the service for trace identification.
    
    Returns:
        A configured OpenTelemetry Tracer.
    """
    global _tracer_provider
    
    # Get configuration from environment
    environment = _get_env("ENVIRONMENT", "production")
    service_version = os.getenv("SERVICE_VERSION", "unknown")
    
    # Parse sampling rate (default to 1.0 if not set or invalid)
    try:
        sampling_rate = float(os.getenv("TRACE_SAMPLING_RATE", "1.0"))
    except ValueError:
        sampling_rate = 1.0
    
    # Create resource
    resource = Resource.create({
        SERVICE_NAME: service_name,
        SERVICE_VERSION: service_version,
        "deployment.environment": environment,
    })
    
    # Create tracer provider with sampling
    _tracer_provider = TracerProvider(
        resource=resource,
        sampler=ParentBasedTraceIdRatio(sampling_rate),
    )
    
    # Configure OTLP span exporter (sends to OTEL collector agent).
    # Endpoint must be a full URL (e.g. http://host:4317); do not strip scheme.
    endpoint = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "http://localhost:4317")
    exporter = OTLPSpanExporter(endpoint=endpoint)
    _tracer_provider.add_span_processor(BatchSpanProcessor(exporter))
    
    # Set as global provider
    trace.set_tracer_provider(_tracer_provider)
    
    return trace.get_tracer(service_name)


def get_trace_id() -> str:
    """Get the current trace ID from the active span.
    
    Returns:
        The trace ID as a hex string, or empty string if no active span.
    """
    span = trace.get_current_span()
    ctx = span.get_span_context()
    if ctx.is_valid:
        return format(ctx.trace_id, "032x")
    return ""


def shutdown_tracer():
    """Shutdown the tracer provider."""
    global _tracer_provider
    if _tracer_provider:
        _tracer_provider.shutdown()
        _tracer_provider = None


def _get_env(key: str, default: str) -> str:
    """Get environment variable with fallback names for ENVIRONMENT."""
    if key == "ENVIRONMENT":
        for k in ["ENVIRONMENT", "ENV", "environment", "env"]:
            v = os.getenv(k)
            if v:
                return v
    
    return os.getenv(key, default)

File: js-sdk/src/observability/tracing/tracer.ts
/**
 * Tracer factory - creates OpenTelemetry tracer with optional Kafka exporter.
 * Aligns with Go NewTracer and Python new_tracer.
 */

import { trace } from '@opentelemetry/api';
import {
  NodeTracerProvider,
  BatchSpanProcessor,
  ParentBasedSampler,
  TraceIdRatioBasedSampler,
} from '@opentelemetry/sdk-trace-node';
import { resourceFromAttributes } from '@opentelemetry/resources';
import { W3CTraceContextPropagator } from '@opentelemetry/core';
import { getEnv, getEnvBool, getEnvNumber } from '../env';
import { KafkaSpanExporter } from './kafka-exporter';
import { getTraceId } from './propagation';

let tracerProvider: NodeTracerProvider | null = null;

export interface TracerResult {
  /** The OpenTelemetry Tracer instance */
  tracer: ReturnType<typeof trace.getTracer>;
  /** Shutdown the tracer provider (call on process exit) */
  shutdown: () => Promise<void>;
}

/**
 * Create a new OpenTelemetry tracer with optional Kafka exporter.
 * In development (ENVIRONMENT=development) or when TRACES_KAFKA_ENABLED=false, no Kafka exporter is added.
 */
export function newTracer(serviceName: string): TracerResult {
  const environment = getEnv('ENVIRONMENT', 'production');
  const serviceVersion = getEnv('SERVICE_VERSION', 'unknown');
  const samplingRate = getEnvNumber('TRACE_SAMPLING_RATE', 1.0);
  const enableKafka = getEnvBool('TRACES_KAFKA_ENABLED', true);
  const brokers = getEnv('KAFKA_BROKERS', '');
  const topic = getEnv('KAFKA_TRACES_TOPIC', 'traces.application');

  const resource = resourceFromAttributes({
    'service.name': serviceName,
    'service.version': serviceVersion,
    'deployment.environment': environment,
  });

  const sampler = new ParentBasedSampler({
    root: new TraceIdRatioBasedSampler(samplingRate),
  });

  const spanProcessors: InstanceType<typeof BatchSpanProcessor>[] = [];

  if (environment !== 'development' && enableKafka && brokers) {
    const kafkaBrokers = brokers.split(',').map((b) => b.trim());
    const exporter = new KafkaSpanExporter({
      serviceName,
      kafkaBrokers,
      topic,
    });
    spanProcessors.push(new BatchSpanProcessor(exporter));
  }

  const provider = new NodeTracerProvider({
    resource,
    sampler,
    spanProcessors: spanProcessors.length > 0 ? spanProcessors : undefined,
  });

  provider.register({
    propagator: new W3CTraceContextPropagator(),
  });

  tracerProvider = provider;

  const otelTracer = trace.getTracer(serviceName, serviceVersion);

  return {
    tracer: otelTracer,
    shutdown: async () => {
      if (tracerProvider) {
        await tracerProvider.shutdown();
        tracerProvider = null;
      }
    },
  };
}

export { getTraceId };
